okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -208,14 +165,23 @@ class ExchangeFinder(
    * If the call's connection exists and is eligible for another exchange, it is returned. If it
    * exists but cannot be used for another exchange, it is closed and this returns null.
    */
-  private fun tryReuseCallConnection(): RealConnection? {
+  private fun tryReuseCallConnection(doExtensiveHealthChecks: Boolean): RealConnection? {
     // This may be mutated by releaseConnectionNoEvents()!
     val candidate = call.connection ?: return null
 
-    var toClose: Socket? = null
-    synchronized(candidate) {
-      if (candidate.noNewExchanges || !sameHostAndPort(candidate.route().address.url)) {
-        toClose = call.releaseConnectionNoEvents()
+    // Make sure this connection is healthy & eligible for new exchanges. If it's no longer needed
+    // then we're on the hook to close it.
+    val healthy = candidate.isHealthy(doExtensiveHealthChecks)
+    val toClose: Socket? = synchronized(candidate) {
```

Comment:
We used to release the call's ownership of the connection when the route didn't match. Now we do it when the route didn’t match or the health check fails.